### PR TITLE
[NNAPI/CoreML EP] Add Onnx opset 14 support

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: [NNAPI/CoreML EP] Add Onnx opset 14 support

**Motivation and Context**
- The only op need special handling (for NNAPI/CoreML EP) in Onnx opset 14 is ReShape with extra attributes "allowzero", detailed changes are here, https://github.com/onnx/onnx/wiki/Logistics-for-ONNX-Release-1.9
- Disable NNAPI for com.microsoft.QLinearConv which is nhwc QLinearConv which is not supported yet

### CODE REVIEW

**Comment by askhade** [comment]
RNN, GRU, LSTM also have changes - layout attribute was added to support batchwise operations

**Comment by gwang-msft** [comment]
> RNN, GRU, LSTM also have changes - layout attribute was added to support batchwise operations

For now NNAPI/CoreML EP does not support RNN/GRN/LSTM, only op change relevant is Reshape

File: onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc [hunk 1]
```diff
@@ -79,7 +79,9 @@ bool ReshapeOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializer
     return false;
   }
 
-  const auto& perm_dims = initializers.at(perm_name)->dims();
+  const auto& shape_tensor = *initializers.at(perm_name);
+  const int64_t* raw_shape = GetTensorInt64Data(shape_tensor);
+  const auto perm_dims = shape_tensor.dims();
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:84]: >auto [](start = 8, length = 4)

can this be `const auto&`?

File: onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc [hunk 2]
```diff
@@ -79,7 +79,9 @@ bool ReshapeOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializer
     return false;
   }
 
-  const auto& perm_dims = initializers.at(perm_name)->dims();
+  const auto& shape_tensor = *initializers.at(perm_name);
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:82]: >shape_tensor [](start = 14, length = 12)

is there a reason that we refer to this initializer as both "perm" and "shape"? it's not so obvious to me

File: onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc [hunk 3]
```diff
@@ -94,6 +96,18 @@ bool ReshapeOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializer
     return false;
   }
 
+  // CoreML reshape does not support 0 as dimension
+  NodeAttrHelper helper(node);
+  bool allow_zero = helper.Get("allowzero ", 0) == 1;
+  if (allow_zero) {
+    for (int64_t i = 0; i < perm_dims[0]; i++) {
+      if (raw_shape[i] == 0) {
+        LOGS_DEFAULT(VERBOSE) << "Reshape doesn't suppport 0 reshape dimension when allowzero is enabled";
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:105]: >suppport [](start = 50, length = 8)

"suppport" -> "support"

Comment (edgchen1) [reply onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:105]: also shows up in op_support_checker.cc

---
In reply to: [606023324](https://github.com/microsoft/onnxruntime/pull/7211#discussion_r606023324) [](ancestors = 606023324)

File: onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc [hunk 4]
```diff
@@ -94,6 +96,18 @@ bool ReshapeOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializer
     return false;
   }
 
+  // CoreML reshape does not support 0 as dimension
+  NodeAttrHelper helper(node);
+  bool allow_zero = helper.Get("allowzero ", 0) == 1;
```
Comment (edgchen1) [inline onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:101]: nit: could be const
similar comment for some bools in the op_support_checker.cc updates

File: onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc (hunk 2 above)
Comment (gwang-msft) [reply onnxruntime/core/providers/coreml/builders/impl/reshape_op_builder.cc:82]: It was copied from NNAPI EP, changed both to perm_tensor, and raw_perm

**Review by edgchen1 — APPROVED** [review]
:shipit:

**Pull request merged**